### DISCUSSION TITLE
Fail if schema used for validation has missing references

### DISCUSSION THREAD

**Discussion by lblatchford** [networknt/json-schema-validator#1116] [Q&A]
Hi - I'm need to be able to validated customer-provided JSON instances against customer-provided JSON schemas.  I have this validation working with the following code:

```
JsonSchemaFactory factory = JsonSchemaFactory.getInstance( VersionFlag.V202012 );

try (InputStream schemaStream = new FileInputStream( schemaFile );
      InputStream instanceStream = new FileInputStream( docFile ))
{
    ObjectMapper objectMapper = new ObjectMapper()
    JsonNode json = objectMapper.readTree( instanceStream );
    JsonSchema schema = factory.getSchema( schemaStream );

    Set<ValidationMessage> errors = schema.validate( json, executionContext -> {   
       executionContext.getExecutionConfig( ).setFormatAssertionsEnabled( true) ;
    } );

    errors.forEach( System.out.println );
}
```

I'd like to prevent validation from being performed for invalid schemas - specifically, I want to be sure that all references are defined.  I haven't figured out a way to accomplish this consistently.  I found a scenario in which InvalidSchemaRefException was thrown if I tried validation against an instance doc of "{}", but this did not work for all schemas with missing references I tried.

For example, the 2020-12 draft schema consists of several files:  schema, meta/applicator, meta/content, meta/core, meta/format-annotation, meta/meta-data, meta/unevaluated and meta/validation.

If I have the 'schema' file present but the meta directory is not, the validation code above still passes.  This is another case of missing references I'd like to catch (though generally this would be with customer schemas versus the draft schema).  I'm guessing validation passed since I have the V202012 metaschema used by default?

How can undefined references be identified whether from missing files or missing definitions in a single schema file?

Suggestions greatly appreciated

**Comment by justin-tay** [comment]
It should already be throwing an exception if references cannot be resolved by default. I'm also not sure why you are mentioning the standard meta schema files as the standard meta schema files are bundled with the library. This library isn't a linter. Typically the best you can do with this library is use the meta schema to validate a schema and attempt to load it to resolve references.

**Reply by lblatchford** [reply]
Thanks for the quick response - I was attempting to use the standard meta schema files because they were a handy example of a schema that has references defined in other files.

I tried a simpler example from the [JSON schema specs doc](https://json-schema.org/understanding-json-schema/structuring), but I think this uncovered a problem in the code in my original post.

I have schema file _address_:
```
{
  "$id": "https://example.com/schemas/address",
  "type": "object",
  "properties": {
    "street_address": { "type": "string" },
    "city": { "type": "string" },
    "state": { "type": "string" }
  },
  "required": ["street_address", "city", "state"]
}
```

Also schema file _customer_:
```
{
  "type": "object",
  "properties": {
    "first_name": { "type": "string" },
    "last_name": { "type": "string" },
    "shipping_address": { "$ref": "https://example.com/schemas/address" },
    "billing_address": { "$ref": "https://example.com/schemas/address" }
  },
  "required": ["first_name", "last_name", "shipping_address", "billing_address"]
}
```

The _customer_ and _address_ schema files are in the same directory.

Validation with my previous code against an instance using the _customer_ schema file is failing with "Failed to load json schema from https://example.com/schemas/address"  "java.net.UnknownHostException:  example.com"

The only input I am given is one or more schema files and an instance document.  I am not provided a URI for each schema, and the directory containing the schemas cannot be on the classpath.

Can I configure the library so that it does not attempt a socket connection to example.com but rather to read the referenced schemas from the  directory?

**Reply by justin-tay** [reply]
See https://github.com/networknt/json-schema-validator/blob/master/doc/schema-retrieval.md

**Reply by lblatchford** [reply]
I was looking at that - it seems to work if I map the $id in the root node of each schema to a file URI for associated schema.  I thought schemas that did not have $id like _customer_ above would be an issue but validation of a _customer_ instance with the id-less schema omitted from the mapping seems to work fine - is this expected?  Thanks again for your quick responses - very much appreciated!
